var/spack/repos/builtin/packages/callpath/package.py

```diff
@@ -25,6 +25,7 @@ class Callpath(CMakePackage):
     depends_on("adept-utils")
     depends_on("mpi")
     depends_on("cmake@2.8:", type="build")
+    conflicts("cmake@4:")
```

Comment:
I think more correct is `^cmake@4:`, but with current syntax `%cmake@4:` should probably be most correct?

In the documentation I put a section on forward and backward compat bounds: https://spack.readthedocs.io/en/latest/packaging_guide.html#specifying-backward-and-forward-compatibility where I recommended `depends_on("cmake@:3", type="build")` for forward compat bounds.

Reply:
is `%` no longer the compiler interface?
Also this is a direct dependency, not a transitive/provider dep, so are you trying to say for this and anything that depends on it? I don't think that's correct unless it exports some feature that is incompatible w/ CMake 4.0.

Reply:
> but as of yesterday %cmake@4: should probably be most correct?

I'd wait to use this syntax for non-compilers until we get the other parser modifications in. Caveat is that I think this will work for `%cmake@X` but will warn for `%cmake +foo`

Reply:
> so are you trying to say for this and anything that depends on it?

No, `conflicts("cmake@4:")` is worse than `conflicts("^cmake@4:")`. The former means it can't be anywhere in the DAG, the latter means it can't be anywhere in the sub-dag of dependencies.

Both are bad :) so I would suggest replace by `depends_on("cmake@:3"[, when="build_system=cmake"], type="build")`.

Reply:
I was actually unaware of the nuances there, thanks for the context. Makes a lot of sense, thanks.

Reply:
This is no longer part of this PR. Resolving.